File: bin/pgsql/dbt-plot-pgsql-table-stats.jl
#!/bin/sh
#
# This file is released under the terms of the Artistic License.
# Please see the file LICENSE, included in this package, for details.
#
# Copyright (C) 2023 Mark Wong
#

#=
usage()
{
	echo "$(basename "$0") is the PostgreSQL table statistics chart generator"
	echo ""
	echo "Usage:"
	echo "  $(basename "$0") [OPTION]"
	echo ""
	echo "Options:"
	echo "  -i CSV      path to CSV file"
	echo "  -n TABLE    table name"
	echo "  -o PATH     output directory"
}

error() {
	echo "ERROR: $*"
	exit 1
}

while getopts "hi:n:o:" opt; do
	case $opt in
	h)
		usage
		exit 1
		;;
	i)
		CSVFILE=$OPTARG
		;;
	n)
		TABLENAME=$OPTARG
		;;
	o)
		OUTDIR=$OPTARG
		mkdir -p "$OUTDIR"
		if [ ! -d "$OUTDIR" ]; then
			echo "ERROR: Failed to create directory $OUTDIR"
			exit 1
		fi
		;;
	\?)
		usage
		exit 1
		;;
	esac
done

if [ "$CSVFILE" = "" ]; then
	error "input CSV file was not specified with -i"
fi

if [ "$OUTDIR" = "" ]; then
	error "output directory was not specified with -o"
fi

if [ "$TABLENAME" = "" ]; then
	error "table name not set, use -n"
fi

exec julia --color=no --startup-file=no "$0" $CSVFILE $OUTDIR $TABLENAME
=#

using CSV
using DataFrames
using VegaLite

function load(params)
    df = CSV.File(params["filename"]) |> DataFrame

    # Convert ctime to elapsed time in minutes, in-place.

    start_time = minimum(df.ctime)
    subset!(df, :relname => ByRow(==(params["relname"])); skipmissing=true)
    transform!(df, :ctime => ByRow(x -> (x - start_time) / 60) => :ctime)

    return df
end

function plot(df, params)
    # Recalculate change between samples.

    for row in reverse(2:nrow(df))
        df.heap_blks_hit[row] =
                df.heap_blks_hit[row] - df.heap_blks_hit[row - 1]
        df.heap_blks_read[row] =
                df.heap_blks_read[row] - df.heap_blks_read[row - 1]
        df.idx_blks_hit[row] =
                df.idx_blks_hit[row] - df.idx_blks_hit[row - 1]
        df.idx_blks_read[row] =
                df.idx_blks_read[row] - df.idx_blks_read[row - 1]
        df.idx_scan[row] = df.idx_scan[row] - df.idx_scan[row - 1]
        df.idx_tup_fetch[row] =
                df.idx_tup_fetch[row] - df.idx_tup_fetch[row - 1]
        df.n_dead_tup[row] = df.n_dead_tup[row] - df.n_dead_tup[row - 1]
        df.n_live_tup[row] = df.n_live_tup[row] - df.n_live_tup[row - 1]
        df.n_tup_del[row] = df.n_tup_del[row] - df.n_tup_del[row - 1]
        df.n_tup_hot_upd[row] =
                df.n_tup_hot_upd[row] - df.n_tup_hot_upd[row - 1]
        df.n_tup_ins[row] = df.n_tup_ins[row] - df.n_tup_ins[row - 1]
        df.n_tup_upd[row] = df.n_tup_upd[row] - df.n_tup_upd[row - 1]
        df.seq_scan[row] = df.seq_scan[row] - df.seq_scan[row - 1]
        df.seq_tup_read[row] = df.seq_tup_read[row] - df.seq_tup_read[row - 1]
        df.tidx_blks_hit[row] =
                df.tidx_blks_hit[row] - df.tidx_blks_hit[row - 1]
        df.tidx_blks_read[row] =
                df.tidx_blks_read[row] - df.tidx_blks_read[row - 1]
        df.toast_blks_hit[row] =
                df.toast_blks_hit[row] - df.toast_blks_hit[row - 1]
        df.toast_blks_read[row] =
                df.toast_blks_read[row] - df.toast_blks_read[row - 1]
    end
    df.heap_blks_hit[1] = 0
    df.heap_blks_read[1] = 0
    df.idx_blks_hit[1] = 0
    df.idx_blks_read[1] = 0
    df.idx_scan[1] = 0
    df.idx_tup_fetch[1] = 0
    df.n_dead_tup[1] = 0
    df.n_live_tup[1] = 0
    df.n_tup_del[1] = 0
    df.n_tup_hot_upd[1] = 0
    df.n_tup_ins[1] = 0
    df.n_tup_upd[1] = 0
    df.seq_scan[1] = 0
    df.seq_tup_read[1] = 0
    df.tidx_blks_hit[1] = 0
    df.tidx_blks_read[1] = 0
    df.toast_blks_hit[1] = 0
    df.toast_blks_read[1] = 0

    plot_table_stat(df, params, "head_blks_hit",
            "Heap Blocks Hit (" * params["relname"] * ")", "Blocks")
    plot_table_stat(df, params, "head_blks_read",
            "Heap Blocks Read (" * params["relname"] * ")", "Blocks")
    plot_table_stat(df, params, "idx_blks_hit",
            "Index Blocks Hit (" * params["relname"] * ")", "Blocks")
    plot_table_stat(df, params, "idx_blks_read",
            "Index Blocks Read (" * params["relname"] * ")", "Blocks")
    plot_table_stat(df, params, "idx_scan",
            "Index Scans (" * params["relname"] * ")", "Index Scans")
    plot_table_stat(df, params, "n_tup_del",
            "Deletes (" * params["relname"] * ")", "Tuples")
    plot_table_stat(df, params, "n_tup_hot_upd",
            "HOT Updates(" * params["relname"] * ")", "Tuples")
    plot_table_stat(df, params, "n_dead_tup",
            "Estimated Number of Dead Rows (" * params["relname"] * ")",
            "Tuples")
    plot_table_stat(df, params, "n_live_tup",
            "Live Rows (" * params["relname"] * ")", "Tuples")
    plot_table_stat(df, params, "n_tup_ins",
            "Inserts (" * params["relname"] * ")", "Tuples")
    plot_table_stat(df, params, "n_tup_upd",
            "Updates (" * params["relname"] * ")", "Tuples")
    plot_table_stat(df, params, "seq_scan",
            "Sequential Scans (" * params["relname"] * ")", "Sequential Scans")
    plot_table_stat(df, params, "seq_tup_read",
            "Tuples Read by Sequential Scans (" * params["relname"] * ")",
            "Tuples")
    plot_table_stat(df, params, "tidx_blks_hit",
            "Toast Index Blocks Hit (" * params["relname"] * ")", "Blocks")
    plot_table_stat(df, params, "tidx_blks_read",
            "Toast Index Blocks Read (" * params["relname"] * ")", "Blocks")
    plot_table_stat(df, params, "toast_blks_hit",
            "Toast Blocks Hit (" * params["relname"] * ")", "Blocks")
    plot_table_stat(df, params, "toast_blks_read",
            "Toast Blocks Read (" * params["relname"] * ")", "Blocks")


    sdf = df[!, [:ctime, :seq_tup_read]]
    rename!(sdf, [:ctime, :tuples])
    insertcols!(sdf, :label => "Sequential Scans")

    ndf = df[!, [:ctime, :idx_tup_fetch]]
    rename!(ndf, [:ctime, :tuples])
    insertcols!(ndf, :label => "Index Scans")
    sdf = vcat(sdf, ndf)

    ndf = df[!, [:ctime, :n_tup_ins]]
    rename!(ndf, [:ctime, :tuples])
    insertcols!(ndf, :label => "Inserts")
    sdf = vcat(sdf, ndf)

    ndf = df[!, [:ctime, :n_tup_upd]]
    rename!(ndf, [:ctime, :tuples])
    insertcols!(ndf, :label => "Updates")
    sdf = vcat(sdf, ndf)

    ndf = df[!, [:ctime, :n_tup_del]]
    rename!(ndf, [:ctime, :tuples])
    insertcols!(ndf, :label => "Deletes")
    sdf = vcat(sdf, ndf)

    ndf = df[!, [:ctime, :n_tup_hot_upd]]
    rename!(ndf, [:ctime, :tuples])
    insertcols!(ndf, :label => "HOT Updates")
    sdf = vcat(sdf, ndf)

    p = sdf |>
        @vlplot(
            title="Tuples (" * params["relname"] * ")",
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :ctime,
                title="Elapsed Time (minutes)",
            },
            y={
                :tuples,
                title="Tuples",
            },
            color={
                :label,
                title="",
            },
        )
    filename = params["outputdir"] * "/table-stat-" * params["relname"] *
            "-tup.png"
    save(filename, p)
end

function plot_table_stat(df, params, field, title, ylabel)
    p = df |>
        @vlplot(
            title=title,
            width=1200,
            height=800,
            mark={
                type="line",
                point=true,
            },
            x={
                :ctime,
                title="Elapsed Time (minutes)",
            },
            y={
                field=field,
                title=ylabel,
            },
            color={
                :schemaname,
                title="Schema",
            },
        )
    filename = params["outputdir"] * "/table-stat-" * params["relname"] * "-" *
            field * ".png"
    save(filename, p)
end

function main()
    params = Dict(
            "filename" => ARGS[1],
            "outputdir" => ARGS[2],
            "relname" => ARGS[3],
    )

    df = load(params)
    plot(df, params)
end

main()

File: bin/pgsql/dbt-plot-pgsql-table-stats.r
#!/bin/sh
#
# This file is released under the terms of the Artistic License.
# Please see the file LICENSE, included in this package, for details.
#
# Copyright (C) 2014      2ndQuadrant, Ltd.
#               2010-2022 Mark Wong
#

usage()
{
	echo "`basename $0` is the PostgreSQL table statistics chart generator"
	echo ""
	echo "Usage:"
	echo "  `basename $0` [OPTION]"
	echo ""
	echo "Options:"
	echo "  -i CSV      path to CSV file"
	echo "  -n TABLE    table name"
	echo "  -o PATH     output directory"
}

error() {
	echo "ERROR: $@"
	exit 1
}

while getopts "hi:n:o:" opt; do
	case $opt in
	h)
		usage
		exit 1
		;;
	i)
		CSVFILE=$OPTARG
		;;
	n)
		TABLENAME=$OPTARG
		;;
	o)
		OUTDIR=$OPTARG
		mkdir -p $OUTDIR
		if [ ! -d "$OUTDIR" ]; then
			echo "ERROR: Failed to create directory $OUTDIR"
			exit 1
		fi
		;;
	\?)
		usage
		exit 1
		;;
	esac
done

if [ "x$CSVFILE" = "x" ]; then
	error "input CSV file was not specified with -i"
fi

if [ "x$OUTDIR" = "x" ]; then
	error "output directory was not specified with -o"
fi

if [ "x$TABLENAME" = "x" ]; then
	error "table name not set, use -n"
fi

R --slave --no-save << __EOF__
df <- read.csv("$CSVFILE", header=T)
df <- subset(df, df\$relname == "${TABLENAME}")
starttime = df[1,]\$ctime
df\$ctime <- (df\$ctime - starttime) / 60

# Calculate rate of change.

tmp <- c(NA, df\$seq_scan[-nrow(df)])
df\$seq_scan <- df\$seq_scan - tmp

tmp <- c(NA, df\$idx_scan[-nrow(df)])
df\$idx_scan <- df\$idx_scan - tmp

tmp <- c(NA, df\$seq_tup_read[-nrow(df)])
df\$seq_tup_read <- df\$seq_tup_read - tmp

tmp <- c(NA, df\$idx_tup_fetch[-nrow(df)])
df\$idx_tup_fetch <- df\$idx_tup_fetch - tmp

tmp <- c(NA, df\$n_tup_ins[-nrow(df)])
df\$n_tup_ins <- df\$n_tup_ins - tmp

tmp <- c(NA, df\$n_tup_upd[-nrow(df)])
df\$n_tup_upd <- df\$n_tup_upd - tmp

tmp <- c(NA, df\$n_tup_del[-nrow(df)])
df\$n_tup_del <- df\$n_tup_del - tmp

tmp <- c(NA, df\$n_tup_hot_upd[-nrow(df)])
df\$n_tup_hot_upd <- df\$n_tup_hot_upd - tmp

#tmp <- c(NA, df\$n_live_tup[-nrow(df)])
#df\$n_live_tup <- df\$n_live_tup - tmp

#tmp <- c(NA, df\$n_dead_tup[-nrow(df)])
#df\$n_dead_tup <- df\$n_dead_tup - tmp

tmp <- c(NA, df\$heap_blks_read[-nrow(df)])
df\$heap_blks_read <- df\$heap_blks_read - tmp

tmp <- c(NA, df\$heap_blks_hit[-nrow(df)])
df\$heap_blks_hit <- df\$heap_blks_hit - tmp

tmp <- c(NA, df\$idx_blks_read[-nrow(df)])
df\$idx_blks_read <- df\$idx_blks_read - tmp

tmp <- c(NA, df\$idx_blks_hit[-nrow(df)])
df\$idx_blks_hit <- df\$idx_blks_hit - tmp

tmp <- c(NA, df\$toast_blks_read[-nrow(df)])
df\$toast_blks_read <- df\$toast_blks_read - tmp

tmp <- c(NA, df\$toast_blks_hit[-nrow(df)])
df\$toast_blks_hit <- df\$toast_blks_hit - tmp

tmp <- c(NA, df\$tidx_blks_read[-nrow(df)])
df\$tidx_blks_read <- df\$tidx_blks_read - tmp

tmp <- c(NA, df\$tidx_blks_hit[-nrow(df)])
df\$tidx_blks_hit <- df\$tidx_blks_hit - tmp

bitmap("${OUTDIR}/table-stat-${TABLENAME}-seq_scan.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$seq_scan, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$seq_scan, na.rm = T)), type = "b", col = rainbow(1),
     main="Sequential Scans (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Sequential Scans")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-seq_tup_read.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$seq_tup_read, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$seq_tup_read, na.rm = T)), type = "b",
     col = rainbow(1), main="Tuples Read by Sequential Scan (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-idx_scan.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$idx_scan, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$idx_scan, na.rm = T)), type = "b", col = rainbow(1),
     main="Index Scans (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Index Scans")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-idx_tup_fetch.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$idx_tup_fetch, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$idx_tup_fetch, na.rm = T)), type = "b",
     col = rainbow(1), main="Tuples Read by Index Scans (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-n_tup_ins.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$n_tup_ins, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$n_tup_ins, na.rm = T)), type = "b", col = rainbow(1),
     main="Inserts (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-n_tup_upd.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$n_tup_upd, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$n_tup_upd, na.rm = T)), type = "b", col = rainbow(1),
     main="Updates (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-n_tup_del.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$n_tup_del, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$n_tup_del, na.rm = T)), type = "b", col = rainbow(1),
     main="Deletes (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-n_tup_hot_upd.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$n_tup_hot_upd, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$n_tup_hot_upd, na.rm = T)), type = "b",
     col = rainbow(1), main="HOT Updates (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-n_live_tup.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$n_live_tup, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$n_live_tup, na.rm = T)), type = "b", col = rainbow(1),
     main="Live Rows (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-n_dead_tup.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$n_dead_tup, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$n_dead_tup, na.rm = T)), type = "b", col = rainbow(1),
     main="Estimated Number of Dead Rows (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-heap_blks_read.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$heap_blks_read, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$heap_blks_read, na.rm = T)), type = "b",
     col = rainbow(1), main="Heap Blocks Read (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Blocks")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-heap_blks_hit.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$heap_blks_hit, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$heap_blks_hit, na.rm = T)), type = "b",
     col = rainbow(1), main="Heap Blocks Hit (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Blocks")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-idx_blks_read.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$idx_blks_read, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$idx_blks_read, na.rm = T)), type = "b",
     col = rainbow(1), main="Index Blocks Read (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Blocks")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-idx_blks_hit.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$idx_blks_hit, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$idx_blks_hit, na.rm = T)), type = "b",
     col = rainbow(1), main="Index Blocks Hit (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Blocks")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-toast_blks_read.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$toast_blks_read, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$toast_blks_read, na.rm = T)), type = "b",
     col = rainbow(1), main="Toast Blocks Read (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Blocks")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-toast_blks_hit.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$toast_blks_hit, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$toast_blks_hit, na.rm = T)), type = "b",
     col = rainbow(1), main="Toast Blocks Hit (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Blocks")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-tidx_blks_read.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$tidx_read, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$tidx_read, na.rm = T)), type = "b",
     col = rainbow(1), main="Toast Index Blocks Read (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Blocks")
grid(col="gray")
invisible(dev.off())

bitmap("${OUTDIR}/table-stat-${TABLENAME}-tidx_blks_hit.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$tidx_blks_hit, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(df\$tidx_blks_hit, na.rm = T)), type = "b",
     col = rainbow(1), main="Toast Index Blocks Hit (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Blocks")
grid(col="gray")
invisible(dev.off())

# Group plots

color = rainbow(2)
pch <- c(1, 2)

bitmap("${OUTDIR}/table-stat-${TABLENAME}-scans.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$seq_scan, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(max(df\$seq_scan, na.rm = T),
                   max(df\$idx_scan, na.rm = T))),
     type = "b", col = rainbow(1), main="Scans on ${TABLENAME}",
     xlab="Elapsed Time (minutes)", ylab="Number of Scans")
points(df\$ctime, df\$idx_scan, type = "b", pch = pch[2], col=color[2])
legend('topright', c("Table Scans", "Index Scans"), pch = pch, col=color)
grid(col="gray")
invisible(dev.off())

color = rainbow(6)
pch <- c(1, 6)

bitmap("${OUTDIR}/table-stat-${TABLENAME}-tup.png",
       type="png16m", units="px", width=1280, height=800, res=150, taa=4,
       gaa=4)
plot(df\$ctime, df\$seq_tup_read, xlim=c(0, max(df\$ctime)),
     ylim=c(0, max(max(df\$seq_tup_read, na.rm = T),
                   max(df\$idx_tup_fetch, na.rm = T),
                   max(df\$n_tup_ins, na.rm = T),
                   max(df\$n_tup_upd, na.rm = T),
                   max(df\$n_tup_del, na.rm = T),
                   max(df\$n_tup_hot_upd, na.rm = T))),
     type = "b", col = rainbow(1),
     main="Tuples (${TABLENAME})",
     xlab="Elapsed Time (minutes)", ylab="Tuples")
points(df\$ctime, df\$idx_tup_fetch, type = "b", pch = pch[2], col=color[2])
points(df\$ctime, df\$n_tup_ins, type = "b", pch = pch[3], col=color[3])
points(df\$ctime, df\$n_tup_upd, type = "b", pch = pch[4], col=color[4])
points(df\$ctime, df\$n_tup_del, type = "b", pch = pch[5], col=color[5])
points(df\$ctime, df\$n_tup_hot_upd, type = "b", pch = pch[6], col=color[6])
legend('topright', c("Sequential Scans", "Index Scans", "Inserts", "Updates",
                     "Deletes", "HOT Updates"), pch = pch, col=color)
grid(col="gray")
invisible(dev.off())
__EOF__
